### ISSUE TITLE
ASM 6.2.1 dependency causes issues on JDK 12 and 13

### ISSUE THREAD

**Issue by lukehutch** [moditect/moditect#108]
Moditect right now depends upon ASM 6.2.1, which causes issues on JDK 12 and 13:

https://issues.apache.org/jira/browse/MJAR-268

Note that it is a *very bad idea* to incorporate a shaded version of ASM into your project now that Java has a 6-month release cadence, because the classfile version number is incremented every 6 months, with each new release, whether or not there were any actual changes to the bytecode format, and by policy, ASM will throw an exception and refuse to parse any classfiles compiled on newer versions of the JDK that it doesn't know about yet (even if it could technically handle the classfiles generated by newer JDKs just fine). So to stay current, a project has to bump the ASM version dependency every 6 months from now on. By incorporating a shaded copy of ASM in Moditect, it's impossible for a dependency of Moditect to override Moditect's ASM dependency in their own `pom.xml`.

Can you please upgrade Moditect to ASM 7.2 and push out a new release?

**Comment by gunnarmorling** [comment]
Hum, hum. Until now I understood it'd be a best practice to shade ASM, so to not be confronted with a new version at runtime (which is what you suggest) whose semantics (new Java language features) you don't understand. Need to read up on this again and better understand recommendations/implications. Probably it's fine to get updates in in our case, as updates to the module info minor grammar seem rare (none so far), and that's the only part we care about.

**Comment by lukehutch** [comment]
Right. FWIW I had to drop the dependency on Moditect for ClassGraph, because the shaded ASM was causing too many issues.

**Comment by gunnarmorling** [comment]
Doh, sorry to hear, and sorry for not getting earlier to this. This project is a bit on the backburner for me atm. I'll still try and see whether I can get rid of the shading and do the update.

**Comment by lukehutch** [comment]
All good. It was simpler in the end just to pre-build the module descriptor, and pull it into the jar before deployment.

**Comment by gunnarmorling** [comment]
> pre-build the module descriptor

Now I'm curious, how are you doing this, if you don't mind me asking?

**Comment by lukehutch** [comment]
I just compiled it by hand in JDK 9 release mode:

https://github.com/classgraph/classgraph/tree/master/src/module-info

(note the need for Placeholder.java)

Then moved the classfile in at the last minute

https://github.com/classgraph/classgraph/blob/master/pom.xml#L324

**Comment by gunnarmorling** [comment]
Gotcha, thx.

>

**Comment by lukehutch** [comment]
Incidentally (if I remember right) it was Animal Sniffer that was failing during build due to the shaded version of ASM in ModiTect, since Animal Sniffer had a dependency on a different version of ASM. You would think the shading would fix that, but somehow it broke. I don't exactly remember the details right now.

**Comment by gunnarmorling** [comment]
Yeah, it's all a bit of a mess. I'm having doubts about the increased class
file version with each release. I hear the reasons for doing it, but it
still seems to put lots of burden onto the eco-system, as nicely proven
here.

>

**Comment by lukehutch** [comment]
Oh, it's a completely braindead policy to bump the ABI version even when there are no semantic changes, but check out just how far the JDK core team dug in on this issue when it was raised on jdk-dev. I lost most of what remaining faith I had in the JDK core team after having been involved in about three or four similarly ridiculous conversations on JDK mailing lists in the last few months.

https://mail.openjdk.java.net/pipermail/jdk-dev/2019-October/003388.html

It's worth reading the whole thread. 

https://mail.openjdk.java.net/pipermail/jdk-dev/2019-October/thread.html

I tried...
